bidict/_base.py

```diff
@@ -184,7 +184,8 @@ def inverse(self) -> 'BidictBase[VT, KT]':
         self._invweak: 't.Optional[weakref.ReferenceType[BidictBase[VT, KT]]]' = None
         # Also store a weak reference back to `instance` on its inverse instance, so that
         # the second `.inverse` access in `bi.inverse.inverse` hits the cached weakref.
-        inv._inv, inv._invweak = None, weakref.ref(self)
+        inv._inv = None
+        inv._invweak = weakref.ref(self)
```

Comment:
> In general, x = 1; y = 2; z = 3 # ... is slower than x, y, z = 1, 2, 3

Are you sure? That's the opposite of what I'd expect, because with the first variant you shouldn't pay the cost of creating an intermediate tuple. I tried some quick benchmarks and also found that the separate assignments are faster. It's a pretty marginal difference, though, so I wouldn't make a decision based on performance unless it's highly performance critical code.

As for the pyright error, I suspect it's a limitation in pyright's bidirectional type inference.

Reply:
> Are you sure? That's the opposite of what I'd expect...

It's the opposite of what I'd expect too, but when I saw this as `#11` in https://youtu.be/YjHsOrOOSuI?t=1216 and reproduced it, it stuck with me ever since. Interesting you're not reproducing it now! I wonder if this changed in more recent versions of Python.

> I wouldn't make a decision based on performance unless it's highly performance critical code.

For sure. (When I said "minor", this is what I should have said.:)

Reply:
> It's the opposite of what I'd expect too, but when I saw this as #11 in https://youtu.be/YjHsOrOOSuI?t=1216 and reproduced it, it stuck with me ever since. Interesting you're not reproducing it now! I wonder if this changed in more recent versions of Python.

(Sorry for going off on this tangent, but now I'm interested :) )

I tried code like `def f(): x1, x2 = 1, 2; return x1, x2` for different numbers of variables (all on 3.6). Unpacking is faster at n = 10, but slower at n = 3 and n = 100. I think the issue is that `(1, 2)` gets turned into a single constant by the compiler, and at n = 10 that apparently makes the whole thing cheaper than repeated assignment would be. But at n = 100 the compiler gives up on generating a single constant tuple and builds it at runtime instead, so repeated assignment is faster again. But in real code you're probably not just assigning a tuple of constant ints, so it's pretty unlikely that you'd run into a situation where it's faster in practice.

Reply:
This is intended behavior in pyright, and it is related to [bidirectional type inference](https://github.com/microsoft/pyright/blob/main/docs/type-inference.md#bidirectional-type-inference-expected-types). Pyright applies bidirectional type inference for assignments if the LHS expression has a declared type. The expression `inv._invweak` does have a declared type, but the expression `inv._inv, inv._invweak` does not. So bidirectional type inference is not applied when you use tuple packing/unpacking.

The other problem here is that you have declared `self._invweak` to have the type `Optional[weakref.ReferenceType[BidictBase[VT, KT]]]', but really it should be declared as `Optional[weakref.ReferenceType[Self]]`. If it were declared as such, then this problem wouldn't exist because the types on the LHS and RHS of the assignment would already match even without the use of bidirectional type inference.

The problem is that the `Self` type was added only recently (for Python 3.11). It is supported in `typing_extensions` and in pyright, but I don't think mypy has support for it yet.

I don't know why mypy accepts the assignment in this case without emitting an error. I don't think it technically should.

Reply:
Thanks for the tip about the [Self type](https://peps.python.org/pep-0673/), I hadn't seen that! I just tested it though and it doesn't currently work with generics. At least, if you try to do something like the following, you get an error about Self not supporting indexing:
```python
class BidictBase(Mapping[KT, VT]):
    def inverse(self) -> Self[VT, KT]:
        ...
```
Without that, I'm not sure how useful Self would be here, even if bidict did only have to support Python 3.11+.